File: CFOnlineShop/CFOnlineShop/Library/CFEditCollectionCell/CFEditCollectionCell.m
//
//  CFEditCollectionCell.m
//  CFOnlineShop
//
//  Created by chenfeng on 2018/7/24.
//  Copyright © 2018年 chenfeng. All rights reserved.
//

#import "CFEditCollectionCell.h"

#define kDeleteBtnWidth 70

@interface CFEditCollectionCell ()<UIGestureRecognizerDelegate>
{
    CGPoint origin;
    BOOL isLeft;
    UIButton *deleteButton;
}

@property (nonatomic, strong) UIPanGestureRecognizer *pan;
@property (nonatomic, strong) UITapGestureRecognizer *tap;

@end

@implementation CFEditCollectionCell

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {

        _status = CFEditCollectionCellStatusWithNormal;
        
        self.contentView.backgroundColor = kWhiteColor;
        
        _pan = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(panGestureDidPan:)];
        _pan.delegate = self;
        
        _tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapGesture:)];
        _tap.delegate = self;
        [self.contentView addGestureRecognizer:_tap];
    }
    return self;
}

- (void)configCollectionCellType:(CFEditCollectionCellType )type
{
    [self.contentView addGestureRecognizer:_pan];
    
    if (type == CFEditCollectionCellTypeWithDelete) {

        if (deleteButton == nil) {
            deleteButton = [UIButton buttonWithType:(UIButtonTypeCustom)];
            deleteButton.frame = CGRectMake(self.mj_w - kDeleteBtnWidth, 0, kDeleteBtnWidth, self.mj_h);
            [deleteButton addTarget:self action:@selector(deleteAction:) forControlEvents:(UIControlEventTouchUpInside)];
            [deleteButton setImage:[UIImage imageNamed:@"delete"] forState:(UIControlStateNormal)];
            deleteButton.backgroundColor = kRedColor;
            [self addSubview:deleteButton];
            [self sendSubviewToBack:deleteButton];
        }
    }
    else if (type == CFEditCollectionCellTypeWithNone)
    {
        //点击手势保留
        [self.contentView removeGestureRecognizer:_pan];
    }
}

- (void)deleteAction:(UIButton *)sender
{
    if (_deleteButtonAction != nil) {
        [self hiddenButtonsWithAnimation];
        _deleteButtonAction(sender);
    }
}

- (void)hiddenButtonsWithAnimation
{
    if (self.contentView.mj_x != 0) {
        [UIView animateWithDuration:0.15 animations:^{
            self.contentView.mj_x = 0;
            } completion:^(BOOL finished) {
                _status = CFEditCollectionCellStatusWithNormal;
        }];
    }
}

- (void)showButtonsWithAnimation
{
    [UIView animateWithDuration:0.15 animations:^{
        self.contentView.mj_x = -kDeleteBtnWidth*1;
        } completion:^(BOOL finished) {
            _status = CFEditCollectionCellStatusWithEdit;
    }];
}

- (void)tapGesture:(UITapGestureRecognizer *)tap
{
    if (_status == CFEditCollectionCellStatusWithEdit) {
        [self hiddenButtonsWithAnimation];
    }
}

- (void)panGestureDidPan:(UIPanGestureRecognizer *)panGesture {
    
    switch (panGesture.state) {
        case UIGestureRecognizerStateBegan:
        {
            origin = [panGesture translationInView:self];
        }
            break;
        case UIGestureRecognizerStateChanged:
        {
            CGPoint translation = [panGesture translationInView:self];
            isLeft = (translation.x < 0);
            
            NSLog(@"panGesture.view.mj_x = %lf",translation.x);
            
            if (isLeft) {
                //多加了30的缓冲偏移量,因为translation.x并不一直是等差变化，滑动速度越快，中间变化量越大
                if (ABS(translation.x) <= kDeleteBtnWidth*1 + 30 &&
                    _status == CFEditCollectionCellStatusWithNormal) {
                    panGesture.view.mj_x = translation.x;
                    
                }
                //左滑未松开然后又向右滑动
                else if (ABS(translation.x) <= 30 &&
                         _status == CFEditCollectionCellStatusWithEdit)
                {
                    panGesture.view.mj_x = - (kDeleteBtnWidth*1 - translation.x);
                }
                NSLog(@"左");
            }
            else
            {
                if (ABS(translation.x) <= kDeleteBtnWidth*1 + 30 &&
                    _status == CFEditCollectionCellStatusWithEdit) {
                    panGesture.view.mj_x = - (kDeleteBtnWidth*1 - translation.x);
                }
                //右滑未松开然后又向左滑动
                else if (ABS(translation.x) <= 30 &&
                         _status == CFEditCollectionCellStatusWithNormal &&
                         panGesture.view.mj_x != 0)
                {
                    panGesture.view.mj_x = translation.x;
                }
                NSLog(@"右");
            }
        }
            break;
        case UIGestureRecognizerStateEnded:
        {
            NSLog(@"end");
            
            if (isLeft) {
                [self showButtonsWithAnimation];
            }
            else
            {
                [self hiddenButtonsWithAnimation];
            }
        }
            break;
        default:
            break;
            
    }
}

- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer {
    
    //self.superview表示collectionView
    //objc_getAssociatedObject和objc_setAssociatedObject类似字典中的key-value形式
    //用self.superview做主对象，用来存放cell子对象（为什么这样用，实在不了解的关联函数的可以百度查一查）
    CFEditCollectionCell *currentCell = objc_getAssociatedObject(self.superview, @"currentCell");
    
    if (gestureRecognizer == _pan &&
        [gestureRecognizer.view isKindOfClass:[UIView class]]) {
        
        CGPoint translation = [_pan translationInView:self];

        objc_setAssociatedObject(self.superview, @"currentCell", self, OBJC_ASSOCIATION_ASSIGN);
        
        //判断滑动
        if (fabs(translation.y) > fabs(translation.x))//表示竖着滑动
        {
            //由于collectionView有边距，触摸到collectionView边距时，这个方法不会生效，所以滑动的时候在外部判断隐藏
            //[currentCell hiddenButtonsWithAnimation];//关闭当前左滑的cell
            return NO;//禁止cell的竖向滑动，使其响应collectionview的上下滚动
        }
        
        if (currentCell != self && currentCell != nil) {
            [currentCell hiddenButtonsWithAnimation];
        }
        
        //表示横着滑动
        return YES;
    }
    else if (gestureRecognizer == _tap &&
             [gestureRecognizer.view isKindOfClass:[UIView class]])
    {
        //判断如果当前没有滑动的cell，则关闭cell内部点击事件，响应外部didSelectItemAtIndexPath事件
        //x坐标不为0时，表示当前cell处于左滑状态
        if (currentCell.status == CFEditCollectionCellStatusWithEdit) {
            [currentCell hiddenButtonsWithAnimation];//关闭当前左滑的cell
            
            return YES;
        }
        else
        {
            return NO;
        }
    }
    else if ([gestureRecognizer.view isKindOfClass:[UICollectionView class]])
    {
        return YES;
    }
    
    return NO;
}

@end

File: SwiftOnlineShopDemo/SwiftOnlineShopDemo/Library/CFEditCollectionCell/CFEditCollectionCell.swift
//
//  CFEditCollectionCell.swift
//  SwiftOnlineShopDemo
//
//  Created by chenfeng on 2019/4/29.
//  Copyright © 2019年 chenfeng. All rights reserved.
//

import UIKit




enum CFEditCollectionCellType {
    case none, deleting
}

enum CFEditCollectionCellStatus {
    case normal, editable
}

class CFEditCollectionCell: UICollectionViewCell,UIGestureRecognizerDelegate {
    
    private var status: CFEditCollectionCellStatus!
    //闭包-类似于oc中的block
    var deleteButtonAction: ((UIButton) ->())?
    
    var origin: CGPoint!
    var isLeft: Bool!
    var deleteButton: UIButton!
    var pan: UIPanGestureRecognizer!
    var tap: UITapGestureRecognizer!
    let deleteBtnWidth: CGFloat = 70
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        status = .normal
        
        self.contentView.backgroundColor = UIColor.white;
        
        tap = UITapGestureRecognizer.init(target: self, action: #selector(tapGesture(tap:)))
        tap.delegate = self;
        self.contentView.addGestureRecognizer(tap)
        
    }
    
    func configCollection(cellType:CFEditCollectionCellType) {
        
        if cellType == .deleting {
            
            pan = UIPanGestureRecognizer.init(target: self, action: #selector(panGestureDidPan(panGesture:)))
            pan.delegate = self;
            self.contentView.addGestureRecognizer(pan)
            
            if deleteButton == nil {
                deleteButton = UIButton(type: .custom);
                deleteButton.frame = CGRect(x: self.mj_w - deleteBtnWidth, y: 0, width: deleteBtnWidth, height: self.mj_h)
                deleteButton.addTarget(self, action: #selector(deleteAction(sender:)), for: .touchUpInside)
                deleteButton.setImage(UIImage(named: "delete"), for: .normal)
                deleteButton.backgroundColor = UIColor.red;
                self.addSubview(deleteButton)
                self.sendSubviewToBack(deleteButton)
            }
        }
        else if cellType == .none
        {

        }
    }
    
    @objc func deleteAction(sender:UIButton) {
        deleteButtonAction?(sender)
    }
    
    func hiddenButtonsWithAnimation() {
        
        if self.contentView.mj_x != 0 {
            UIView.animate(withDuration: 0.15, animations: {
                self.contentView.mj_x = 0
            }) { (Bool) in
                self.status = .normal
            }
        }
    }
    
    func showButtonsWithAnimation() {
        UIView.animate(withDuration: 0.15, animations: {
            self.contentView.mj_x = -self.deleteBtnWidth * 1;
        }) { (Bool) in
            self.status = .editable
        }
    }
    
    @objc func tapGesture(tap:UITapGestureRecognizer) {
        if status == .editable {
            hiddenButtonsWithAnimation()
        }
    }
    
    @objc func panGestureDidPan(panGesture:UIPanGestureRecognizer) {
        
        switch panGesture.state
        {
            case .began:
                origin = panGesture.translation(in: self)
            
            case .changed:
                let translation = panGesture.translation(in: self)
                isLeft = (translation.x < 0)
                print("panGesture.view.mj_x =\(translation.x)")
                if isLeft
                {
                    //多加了30的缓冲偏移量,因为translation.x并不一直是等差变化，滑动速度越快，中间变化量越大
                    if abs(translation.x) <= deleteBtnWidth*1 + 30 &&
                        status == .normal {
                        panGesture.view?.mj_x = translation.x;
                        
                    }
                        //左滑未松开然后又向左滑动
                    else if abs(translation.x) <= 30 &&
                        status == .editable
                    {
                        panGesture.view?.mj_x = -(deleteBtnWidth*1 - translation.x);
                    }
                    print("左");
                }
                else
                {
                    if abs(translation.x) <= deleteBtnWidth*1 + 30 &&
                        status == .editable {
                        panGesture.view?.mj_x = -(deleteBtnWidth*1 - translation.x);
                    }
                        //左滑滑未松开然后又向右滑动
                    else if abs(translation.x) <= 30 &&
                        status == .normal &&
                        panGesture.view?.mj_x != 0
                    {
                        panGesture.view?.mj_x = translation.x
                    }
                    print("右");
                }
            
            
            case .ended:
            
                print("end");
                
                if isLeft {
                    showButtonsWithAnimation()
                }
                else
                {
                    hiddenButtonsWithAnimation()
                }
            
            default:
                print("其他");
        }
        
    }
    
    override func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {

        //let currentCell:CFEditCollectionCell? = (self.superview as! UICollectionView).currentCell as? CFEditCollectionCell

        let currentCell:CFEditCollectionCell? = objc_getAssociatedObject((self.superview as! UICollectionView), &currentCellForKey) as? CFEditCollectionCell
        
        if gestureRecognizer == pan &&
            (gestureRecognizer.view?.isKind(of: UIView.classForCoder()))! {
            
            let translation = pan.translation(in: self)
            
            //关联对象
            objc_setAssociatedObject((self.superview as! UICollectionView), &currentCellForKey, self, objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
            
            //(self.superview as! UICollectionView).currentCell = self;

            //判断滑动
            if abs(translation.y) > abs(translation.x)//表示竖着滑动
            {
                //由于collectionView有边距，触摸到collectionView边距时，这个方法不会生效，所以滑动的时候在外部判断隐藏
                //[currentCell hiddenButtonsWithAnimation];//关闭当前左滑的cell
                return false;//禁止cell的竖向滑动，使其响应collectionview的上下滚动
            }
            
            if currentCell != self {
                currentCell?.hiddenButtonsWithAnimation()
            }
            
            //表示横着滑动
            return true;
        }
        else if gestureRecognizer == tap &&
            (gestureRecognizer.view?.isKind(of: UIView.classForCoder()))!
        {
            //判断如果当前没有滑动的cell，则关闭cell内部点击事件，响应外部didSelectItemAtIndexPath事件
            //x坐标不为0时，表示当前cell处于左滑状态
            if currentCell?.status == .editable {
                currentCell?.hiddenButtonsWithAnimation()//关闭当前左滑的cell
                
                return true;
            }
            else
            {
                return false;
            }
        }
        else if (gestureRecognizer.view?.isKind(of: UICollectionView.classForCoder()))!
        {
            return true
        }
        
        return false
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}
